omise/omise.php

```diff
@@ -290,6 +331,32 @@ public function install()
         return true;
     }
 
+    /**
+     * A PrestaShop version specific version of 'display'
+     *
+     * @return see parent 'display' method
+     */
+    protected function versionSpecificDisplay($file, $template)
+    {
+        $versionedTemplate = PRESTASHOP_VERSION_VIEW_PATH . $template;
+        return $this->display($file, $versionedTemplate);
+    }    
+
+    /**
+     * Register/Unregister all hooks for the module
+     *
+     * @return bool
+     */
+    protected function registerHooks($isRegister = true)
+    {
+        $res = true;
+        foreach (explode(',', PRESTASHOP_PAYMENTMODULE_HOOKS) as $hook) {
+            $res = $res && $this->{$isRegister ? "registerHook" : "unregisterHook"}($hook);
```

Comment:
@jonrandy why don't we separate this method to `registerHooks` and `unregisterHooks` 🤔

Reply:
I mean, it's strange that `registerHooks` method needed to decide whether executing `registerHook` or `unregisterHook` depending on `isRegister` is true or false.

Reply:
Both methods would be all but identical - just the name of the method called would change. Pointless. DRY

Reply:
Maybe, but it adds an unnecessary method, and repeated code. Or a tiny convenience method:

```php
function unregisterHooks() {
   $this->registerHooks(false);
}
```

In my mind, a total waste time, and less efficient - 2 function calls instead of 1

Reply:
Another way would be to pass `$this->unregisterHook` or `$this->reigisterHook` into a function `modifyHooks` which would loop through the hooks and run the passed function on them

Reply:
For the above, how about
```php
/**
 * Register/Unregister all hooks for the module
 *
 * @param  string $mode  Either 'registerHook' or 'unregisterHook'
 *
 * @return bool
 */
protected toggleModuleHookRegistration($mode) {
    foreach (explode(',', PRESTASHOP_PAYMENTMODULE_HOOKS) as $hook) {
        if (! $this->{$mode}($hook)) return false;
    }

    return true;
}

/**
 * Register all hooks for the module
 *
 * @return bool
 */
protected function registerHooks()
{
    return $this->toggleModuleHookRegistration("registerHook");
}

/**
 * Unregister all hooks for the module
 *
 * @return bool
 */
protected function unregisterHooks()
{
    return $this->toggleModuleHookRegistration("unregisterHook");
}
```

This introduces on both readability and DRY, and code is a bit more explicit.

We might able to make it even more better by decoupling `PRESTASHOP_PAYMENTMODULE_HOOKS` out from the `toggleModuleHookRegistration` method and passing the module list into `registerHooks($modules)` and `unregisterHooks($modules)` instead (not necessary but it might help improving on the testability part as well, since there we decoupling that constant out from the method.)

_note, just idea. I didn't test if this way work_

Reply:
Understand that few more lines or methods aren't really necessary as the program is still function properly.
But if there is a way to improve its readability and it won't hurt anything, that would really helpful for any developers that are going to maintain or improve this code after us.

```php
return parent::uninstall() && $this->registerHooks(false);
```

This really doesn't sound right, as you can `unregister` hooks by passing `false` at the first argument of the `registerHook` method.

Don't think this thing is pointless thou, let's meet half-way 😉

Reply:
What you suggested is essentially identical to what I suggested (i.e. passing a function into a method that iterates over the hooknames and applies that function on them) except you are passing a method name instead of a function. The ultimate flexibility would obviously be gained by passing in a function, although I've never attempted that in PHP and some quick reading suggests it might be more trouble than it's worth, so passing in a string with a named method is the next best thing. I planned to do this tomorrow anyway. I'll probably experiment with the passing a function method first though as I'm curious how this is done in PHP :stuck_out_tongue: 

As for passing in the hook names, we could, but it is essentially pointless as the hooks are 'fixed' for the plugin - one set for each PrestaShop version. They don't change, and the only time anything is done with them is during the install/uninstall, where all hooks are registered/unregistered. I would change the method if a situation occurred that needed that change. As it stands, no such situation exists so there is no reason to change it. Coding for every possible eventuality is an exercise in futility. The method is as flexible as it needs to be for now.

As for testability - I've never written a test for any 'real' code in my entire life, I've only written tests as part of 'learning how to do automated testing' workshops. As far as I'm concerned, it is a simple method, and I _know_ it works. If I change it, and it breaks - I will see that when I am testing the plugin manually, and I will fix it before putting it live. Writing an automated test is unnecessary. That's how I see it anyway

Reply:
@jonrandy go for it 👍
Let me know if you want me to review the code again after finish experimenting that passing function thing :D